// @atlaskit/button/dist/es5/styled/themeDefinitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var theme_1 = require("@atlaskit/theme");
/**
 * Convert a hex colour code to RGBA.
 * @param {String} hex Hex colour code.
 * @param {Number} alpha Optional alpha value (defaults to 1).
 *
 */
var hex2rgba = function (hex, alpha) {
    if (alpha === void 0) { alpha = 1; }
    if (/^#([A-Fa-f0-9]{3}){1,2}$/.test(hex)) {
        var colorArr = hex.substring(1).split('');
        if (colorArr.length === 3) {
            colorArr = [
                colorArr[0],
                colorArr[0],
                colorArr[1],
                colorArr[1],
                colorArr[2],
                colorArr[2],
            ];
        }
        var color = "0x" + colorArr.join('');
        // FIXME: `>>` operand can validly take a string value
        var r = (color >> 16) & 255;
        var g = (color >> 8) & 255;
        var b = color & 255;
        return "rgba(" + [r, g, b].join(',') + ", " + alpha + ")";
    }
    throw new Error('Bad Hex');
};
/* eslint-enable no-bitwise */
exports.default = {
    // Fallbacks
    fallbacks: {
        background: theme_1.themed({ light: theme_1.colors.N20A, dark: theme_1.colors.DN70 }),
        color: theme_1.themed({ light: theme_1.colors.N400, dark: theme_1.colors.DN400 }),
        textDecoration: 'none',
    },
    // Themes
    theme: {
        // Default appearance
        default: {
            background: {
                default: theme_1.themed({ light: theme_1.colors.N20A, dark: theme_1.colors.DN70 }),
                hover: theme_1.themed({ light: theme_1.colors.N30A, dark: theme_1.colors.DN60 }),
                active: theme_1.themed({ light: hex2rgba(theme_1.colors.B75, 0.6), dark: theme_1.colors.B75 }),
                disabled: theme_1.themed({ light: theme_1.colors.N20A, dark: theme_1.colors.DN70 }),
                selected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.DN0 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.DN0 }),
            },
            boxShadowColor: {
                focus: theme_1.themed({ light: hex2rgba(theme_1.colors.B200, 0.6), dark: theme_1.colors.B75 }),
                focusSelected: theme_1.themed({
                    light: hex2rgba(theme_1.colors.B200, 0.6),
                    dark: theme_1.colors.B75,
                }),
            },
            color: {
                default: theme_1.themed({ light: theme_1.colors.N400, dark: theme_1.colors.DN400 }),
                active: theme_1.themed({ light: theme_1.colors.B400, dark: theme_1.colors.B400 }),
                disabled: theme_1.themed({ light: theme_1.colors.N70, dark: theme_1.colors.DN30 }),
                selected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
            },
        },
        // Primary appearance
        primary: {
            background: {
                default: theme_1.themed({ light: theme_1.colors.B400, dark: theme_1.colors.B100 }),
                hover: theme_1.themed({ light: theme_1.colors.B300, dark: theme_1.colors.B75 }),
                active: theme_1.themed({ light: theme_1.colors.B500, dark: theme_1.colors.B200 }),
                disabled: theme_1.themed({ light: theme_1.colors.N20A, dark: theme_1.colors.DN70 }),
                selected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.DN0 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.DN0 }),
            },
            boxShadowColor: {
                focus: theme_1.themed({ light: hex2rgba(theme_1.colors.B200, 0.6), dark: theme_1.colors.B75 }),
                focusSelected: theme_1.themed({
                    light: hex2rgba(theme_1.colors.B200, 0.6),
                    dark: theme_1.colors.B75,
                }),
            },
            color: {
                default: theme_1.themed({ light: theme_1.colors.N0, dark: theme_1.colors.DN30 }),
                disabled: theme_1.themed({ light: theme_1.colors.N70, dark: theme_1.colors.DN30 }),
                selected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
            },
        },
        // Warning appearance
        warning: {
            background: {
                default: theme_1.themed({ light: theme_1.colors.Y300, dark: theme_1.colors.Y300 }),
                hover: theme_1.themed({ light: theme_1.colors.Y200, dark: theme_1.colors.Y200 }),
                active: theme_1.themed({ light: theme_1.colors.Y400, dark: theme_1.colors.Y400 }),
                disabled: theme_1.themed({ light: theme_1.colors.N20A, dark: theme_1.colors.DN70 }),
                selected: theme_1.themed({ light: theme_1.colors.Y400, dark: theme_1.colors.Y400 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.Y400, dark: theme_1.colors.Y400 }),
            },
            boxShadowColor: {
                focus: theme_1.themed({ light: theme_1.colors.Y500, dark: theme_1.colors.Y500 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.Y500, dark: theme_1.colors.Y500 }),
            },
            color: {
                default: theme_1.themed({ light: theme_1.colors.N800, dark: theme_1.colors.N800 }),
                disabled: theme_1.themed({ light: theme_1.colors.N70, dark: theme_1.colors.DN30 }),
                selected: theme_1.themed({ light: theme_1.colors.N800, dark: theme_1.colors.N800 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N800, dark: theme_1.colors.N800 }),
            },
        },
        // Danger appearance
        danger: {
            background: {
                default: theme_1.themed({ light: theme_1.colors.R400, dark: theme_1.colors.R400 }),
                hover: theme_1.themed({ light: theme_1.colors.R300, dark: theme_1.colors.R300 }),
                active: theme_1.themed({ light: theme_1.colors.R500, dark: theme_1.colors.R500 }),
                disabled: theme_1.themed({ light: theme_1.colors.N20A, dark: theme_1.colors.DN70 }),
                selected: theme_1.themed({ light: theme_1.colors.R500, dark: theme_1.colors.R500 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.R500, dark: theme_1.colors.R500 }),
            },
            boxShadowColor: {
                focus: theme_1.themed({ light: theme_1.colors.R100, dark: theme_1.colors.R100 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.R100, dark: theme_1.colors.R100 }),
            },
            color: {
                default: theme_1.themed({ light: theme_1.colors.N0, dark: theme_1.colors.N0 }),
                disabled: theme_1.themed({ light: theme_1.colors.N70, dark: theme_1.colors.DN30 }),
                selected: theme_1.themed({ light: theme_1.colors.N0, dark: theme_1.colors.N0 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N0, dark: theme_1.colors.N0 }),
            },
        },
        // Help appearance
        help: {
            background: {
                default: theme_1.themed({ light: theme_1.colors.P400, dark: theme_1.colors.P400 }),
                hover: theme_1.themed({ light: theme_1.colors.P200, dark: theme_1.colors.P200 }),
                active: theme_1.themed({ light: theme_1.colors.P500, dark: theme_1.colors.P500 }),
                disabled: theme_1.themed({ light: theme_1.colors.N20A, dark: theme_1.colors.DN70 }),
                selected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.DN0 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.R500, dark: theme_1.colors.R500 }),
            },
            boxShadowColor: {
                focus: theme_1.themed({ light: theme_1.colors.P100, dark: theme_1.colors.P100 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.P100, dark: theme_1.colors.P100 }),
            },
            color: {
                default: theme_1.themed({ light: theme_1.colors.N0, dark: theme_1.colors.N0 }),
                disabled: theme_1.themed({ light: theme_1.colors.N70, dark: theme_1.colors.DN30 }),
                selected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N0, dark: theme_1.colors.N0 }),
            },
        },
        // Link appearance
        link: {
            background: {
                default: theme_1.themed({ light: 'none', dark: 'none' }),
                selected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.N20 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.N20 }),
            },
            boxShadowColor: {
                focus: theme_1.themed({ light: hex2rgba(theme_1.colors.B200, 0.6), dark: theme_1.colors.B75 }),
                focusSelected: theme_1.themed({
                    light: hex2rgba(theme_1.colors.B200, 0.6),
                    dark: theme_1.colors.B75,
                }),
            },
            color: {
                default: theme_1.themed({ light: theme_1.colors.B400, dark: theme_1.colors.B100 }),
                hover: theme_1.themed({ light: theme_1.colors.B300, dark: theme_1.colors.B75 }),
                active: theme_1.themed({ light: theme_1.colors.B500, dark: theme_1.colors.B200 }),
                disabled: theme_1.themed({ light: theme_1.colors.N70, dark: theme_1.colors.DN100 }),
                selected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.N700 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.N700 }),
            },
            textDecoration: {
                hover: 'underline',
            },
        },
        // Subtle appearance
        subtle: {
            background: {
                default: theme_1.themed({ light: 'none', dark: 'none' }),
                hover: theme_1.themed({ light: theme_1.colors.N30A, dark: theme_1.colors.DN60 }),
                active: theme_1.themed({ light: hex2rgba(theme_1.colors.B75, 0.6), dark: theme_1.colors.B75 }),
                disabled: theme_1.themed({ light: 'none', dark: 'none' }),
                selected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.DN0 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.DN0 }),
            },
            boxShadowColor: {
                focus: theme_1.themed({ light: hex2rgba(theme_1.colors.B200, 0.6), dark: theme_1.colors.B75 }),
                focusSelected: theme_1.themed({
                    light: hex2rgba(theme_1.colors.B200, 0.6),
                    dark: theme_1.colors.B75,
                }),
            },
            color: {
                default: theme_1.themed({ light: theme_1.colors.N400, dark: theme_1.colors.DN400 }),
                active: theme_1.themed({ light: theme_1.colors.B400, dark: theme_1.colors.B400 }),
                disabled: theme_1.themed({ light: theme_1.colors.N70, dark: theme_1.colors.DN100 }),
                selected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
            },
        },
        // Subtle Link appearance
        'subtle-link': {
            background: {
                default: theme_1.themed({ light: 'none', dark: 'none' }),
                selected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.N20 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N700, dark: theme_1.colors.N20 }),
            },
            boxShadowColor: {
                focus: theme_1.themed({ light: hex2rgba(theme_1.colors.B200, 0.6), dark: theme_1.colors.B75 }),
                focusSelected: theme_1.themed({
                    light: hex2rgba(theme_1.colors.B200, 0.6),
                    dark: theme_1.colors.B75,
                }),
            },
            color: {
                default: theme_1.themed({ light: theme_1.colors.N200, dark: theme_1.colors.DN400 }),
                hover: theme_1.themed({ light: theme_1.colors.N90, dark: theme_1.colors.B50 }),
                active: theme_1.themed({ light: theme_1.colors.N400, dark: theme_1.colors.DN300 }),
                disabled: theme_1.themed({ light: theme_1.colors.N70, dark: theme_1.colors.DN100 }),
                selected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
                focusSelected: theme_1.themed({ light: theme_1.colors.N20, dark: theme_1.colors.DN400 }),
            },
            textDecoration: {
                hover: 'underline',
            },
        },
    },
};
