// @atlaskit/tag/dist/cjs/Content/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styled = require("./styled");

var Content =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(Content, _Component);

  function Content() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Content);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Content)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "getLinkComponent", function () {
      var _this$props = _this.props,
          linkComponent = _this$props.linkComponent,
          href = _this$props.href;
      if (!href) return null;
      if (linkComponent) return (0, _styledComponents.default)(linkComponent).withConfig({
        displayName: "Content",
        componentId: "nijpvw-0"
      })(["\n        ", ";\n      "], _styled.linkStyles);
      return _styled.Link;
    });
    return _this;
  }

  (0, _createClass2.default)(Content, [{
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          children = _this$props2.children,
          href = _this$props2.href,
          isFocused = _this$props2.isFocused,
          isRemovable = _this$props2.isRemovable,
          markedForRemoval = _this$props2.markedForRemoval,
          color = _this$props2.color;
      var styledProps = {
        isFocused: isFocused,
        isRemovable: isRemovable,
        markedForRemoval: markedForRemoval,
        color: color
      };
      var LinkComponent = this.getLinkComponent();
      return href && LinkComponent ? _react.default.createElement(LinkComponent, (0, _extends2.default)({}, styledProps, {
        href: href,
        tabIndex: "-1"
      }), children) : _react.default.createElement(_styled.Text, styledProps, children);
    }
  }]);
  return Content;
}(_react.Component);

exports.default = Content;
