// @atlaskit/inline-message/dist/cjs/constants.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.typesMapping = exports.itemSpacing = void 0;

var _theme = require("@atlaskit/theme");

var _warning = _interopRequireDefault(require("@atlaskit/icon/glyph/warning"));

var _checkCircle = _interopRequireDefault(require("@atlaskit/icon/glyph/check-circle"));

var _info = _interopRequireDefault(require("@atlaskit/icon/glyph/info"));

var _error = _interopRequireDefault(require("@atlaskit/icon/glyph/error"));

var itemSpacing = (0, _theme.gridSize)() / 2;
exports.itemSpacing = itemSpacing;
var typesMapping = {
  connectivity: {
    icon: _warning.default,
    iconSize: 'medium'
  },
  confirmation: {
    icon: _checkCircle.default,
    iconSize: 'medium'
  },
  info: {
    icon: _info.default,
    iconSize: 'medium'
  },
  warning: {
    icon: _warning.default,
    iconSize: 'medium'
  },
  error: {
    icon: _error.default,
    iconSize: 'medium'
  }
};
exports.typesMapping = typesMapping;
