// @atlaskit/flag/dist/es2019/components/FlagActions/index.js
import React, { Component } from 'react';
import uuid from 'uuid/v1';
import Button from '@atlaskit/button';
import { DEFAULT_APPEARANCE } from '../../constants';
import { actionButtonStyles, getPseudos } from '../../theme';
import Container, { Action } from './styledFlagActions';
export default class FlagActions extends Component {
    constructor() {
        super(...arguments);
        this.getUniqueId = (prefix) => `${prefix}-${uuid()}`;
    }
    render() {
        const { actions, appearance, linkComponent } = this.props;
        const isBold = appearance !== DEFAULT_APPEARANCE;
        if (!actions.length)
            return null;
        return (React.createElement(Container, { appearance: appearance }, actions.map((action, index) => (React.createElement(Action, { key: this.getUniqueId('flag-action'), hasDivider: !!index, useMidDot: !isBold, "data-testid": action.testId },
            React.createElement(Button, { onClick: action.onClick, href: action.href, target: action.target, 
                // This is very much a hack
                // This should be tidied up when the appearance prop
                // of flag is aligned with other appearance props.
                appearance: appearance === 'normal'
                    ? 'link'
                    : appearance, component: linkComponent, spacing: "compact", "data-testid": action.testId, theme: (adgTheme, themeProps) => {
                    const { buttonStyles, ...rest } = adgTheme(themeProps);
                    return {
                        buttonStyles: {
                            ...buttonStyles,
                            ...actionButtonStyles({
                                appearance,
                                mode: themeProps.mode,
                            }),
                            ...getPseudos({ appearance, mode: themeProps.mode }),
                        },
                        ...rest,
                    };
                } }, action.content))))));
    }
}
// eslint-disable-line react/sort-comp
FlagActions.defaultProps = {
    appearance: DEFAULT_APPEARANCE,
    actions: [],
};
