// @atlaskit/droplist/dist/cjs/theme/item-theme.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _item = require("@atlaskit/item");

var _theme = require("@atlaskit/theme");

var dropdownPadding = {
  bottom: _theme.gridSize,
  left: _theme.math.multiply(_theme.gridSize, 1.5),
  right: _theme.math.multiply(_theme.gridSize, 1.5),
  top: _theme.gridSize
};
var droplistItemTheme = {
  padding: {
    default: dropdownPadding,
    compact: dropdownPadding
  },
  borderRadius: function borderRadius() {
    return 0;
  },
  default: {
    background: (0, _theme.themed)({
      light: _theme.colors.N0,
      dark: _theme.colors.DN50
    }),
    text: (0, _theme.themed)({
      light: _theme.colors.N800,
      dark: _theme.colors.DN600
    }),
    secondaryText: (0, _theme.themed)({
      light: _theme.colors.N200,
      dark: _theme.colors.DN300
    })
  },
  hover: {
    background: (0, _theme.themed)({
      light: _theme.colors.N20,
      dark: _theme.colors.DN70
    }),
    text: (0, _theme.themed)({
      light: _theme.colors.N800,
      dark: _theme.colors.DN600
    }),
    secondaryText: (0, _theme.themed)({
      light: _theme.colors.N200,
      dark: _theme.colors.DN300
    })
  },
  active: {
    background: (0, _theme.themed)({
      light: _theme.colors.B75,
      dark: _theme.colors.B75
    }),
    text: (0, _theme.themed)({
      light: _theme.colors.N800,
      dark: _theme.colors.B400
    }),
    secondaryText: (0, _theme.themed)({
      light: _theme.colors.N200,
      dark: _theme.colors.DN300
    })
  },
  selected: {
    background: 'transparent',
    text: (0, _theme.themed)({
      light: _theme.colors.N800,
      dark: _theme.colors.DN600
    }),
    secondaryText: (0, _theme.themed)({
      light: _theme.colors.N200,
      dark: _theme.colors.DN300
    })
  },
  disabled: {
    background: 'transparent',
    text: (0, _theme.themed)({
      light: _theme.colors.N70,
      dark: _theme.colors.DN80
    }),
    secondaryText: (0, _theme.themed)({
      light: _theme.colors.N50,
      dark: _theme.colors.DN70
    })
  },
  focus: {
    outline: (0, _theme.themed)({
      light: _theme.colors.B100,
      dark: _theme.colors.B75
    })
  }
};

var _default = (0, _defineProperty2.default)({}, _item.itemThemeNamespace, droplistItemTheme);

exports.default = _default;
