// @atlaskit/spinner/dist/cjs/Spinner/styledSvg.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.svgStyles = exports.getStrokeColor = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _theme = require("@atlaskit/theme");

var _constants = require("./constants");

var getStrokeWidth = function getStrokeWidth(size) {
  return Math.round(size / 10);
};

var getStrokeCircumference = function getStrokeCircumference(size) {
  var strokeWidth = getStrokeWidth(size);
  var strokeRadius = size / 2 - strokeWidth / 2;
  return Math.PI * strokeRadius * 2;
};
/* Define keyframes statically to prevent a perfomance issue in styled components v1 where the keyframes function
 * does not cache previous values resulting in each spinner injecting the same keyframe definition
 * in the DOM.
 * This can be reverted to dynamic keyframes when we upgrade to styled components v2
 */


var keyframeNames = {
  noop: (0, _styledComponents.keyframes)(["\n    from { opacity: 0; }\n    to { opacity: 0; }\n  "]),
  rotate: (0, _styledComponents.keyframes)(["\n    to { transform: rotate(360deg); }\n  "]),
  enterOpacity: (0, _styledComponents.keyframes)(["\n    from { opacity: 0; }\n    to { opacity: 1; }\n  "]),
  smallEnterStroke: (0, _styledComponents.keyframes)(["\n    from { stroke-dashoffset: ", "px; }\n    to { stroke-dashoffset: ", "px; }\n  "], getStrokeCircumference(_constants.SIZES_MAP.small), getStrokeCircumference(_constants.SIZES_MAP.small) * 0.8),
  mediumEnterStroke: (0, _styledComponents.keyframes)(["\n    from { stroke-dashoffset: ", "px; }\n    to { stroke-dashoffset: ", "px; }\n  "], getStrokeCircumference(_constants.SIZES_MAP.medium), getStrokeCircumference(_constants.SIZES_MAP.medium) * 0.8),
  largeEnterStroke: (0, _styledComponents.keyframes)(["\n    from { stroke-dashoffset: ", "px; }\n    to { stroke-dashoffset: ", "px; }\n  "], getStrokeCircumference(_constants.SIZES_MAP.large), getStrokeCircumference(_constants.SIZES_MAP.large) * 0.8),
  xlargeEnterStroke: (0, _styledComponents.keyframes)(["\n    from { stroke-dashoffset: ", "px; }\n    to { stroke-dashoffset: ", "px; }\n  "], getStrokeCircumference(_constants.SIZES_MAP.xlarge), getStrokeCircumference(_constants.SIZES_MAP.xlarge) * 0.8)
};
/* If a standard size is used, we can use one of our statically defined keyframes, otherwise
 * we're forced to dynamically create the keyframe and incur a performance cost.
 */

var getEnterStrokeKeyframe = function getEnterStrokeKeyframe(size) {
  var standardSizeName = Object.keys(_constants.SIZES_MAP).find(function (sizeName) {
    return size === _constants.SIZES_MAP[sizeName];
  });

  if (standardSizeName) {
    return keyframeNames["".concat(standardSizeName, "EnterStroke")];
  }

  var circumference = getStrokeCircumference(size);
  return (0, _styledComponents.keyframes)(["\n    from { stroke-dashoffset: ", "px; }\n    to { stroke-dashoffset: ", "px; }\n  "], circumference, circumference * 0.8);
};

var spinnerColor = (0, _theme.themed)({
  light: _theme.colors.N500,
  dark: _theme.colors.N0
});
var spinnerColorInverted = (0, _theme.themed)({
  light: _theme.colors.N0,
  dark: _theme.colors.N0
});

var getStrokeColor = function getStrokeColor(_ref // $FlowFixMe - theme is not found in props
) {
  var invertColor = _ref.invertColor,
      props = (0, _objectWithoutProperties2.default)(_ref, ["invertColor"]);
  return invertColor ? spinnerColorInverted(props) : spinnerColor(props);
};

exports.getStrokeColor = getStrokeColor;
var svgStyles = (0, _styledComponents.css)(["\n  ", ";\n"], function (props) {
  var circumference = getStrokeCircumference(props.size);
  var idleRotation = "0.86s cubic-bezier(0.4, 0.15, 0.6, 0.85) infinite ".concat(keyframeNames.rotate);
  var spinUpStroke = "0.8s ease-in-out ".concat(getEnterStrokeKeyframe(props.size));
  var spinUpOpacity = "0.2s ease-in-out ".concat(keyframeNames.enterOpacity);
  var activeAnimations = [idleRotation];

  if (props.phase === 'ENTER') {
    activeAnimations.push(spinUpStroke, spinUpOpacity);
  }

  return (0, _styledComponents.css)(["\n      animation: ", ";\n      fill: none;\n      stroke: ", ";\n      stroke-dasharray: ", "px;\n      stroke-dashoffset: ", "px;\n      stroke-linecap: round;\n      stroke-width: ", "px;\n      transform-origin: center;\n    "], activeAnimations.join(', '), getStrokeColor, circumference, circumference * 0.8, getStrokeWidth(props.size));
});
exports.svgStyles = svgStyles;

var Svg = _styledComponents.default.svg.withConfig({
  displayName: "styledSvg__Svg",
  componentId: "y2l69q-0"
})(["\n  ", ";\n"], svgStyles);

Svg.displayName = 'SpinnerSvg';
var _default = Svg;
exports.default = _default;
