// @atlaskit/item/dist/cjs/util/theme.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getThemeStyle = exports.themeWithKeys = exports.defaultTheme = exports.gridSize = exports.compactLineHeight = exports.compactSmallFontSize = exports.themeNamespace = void 0;

var _theme = require("@atlaskit/theme");

var themeNamespace = '@atlaskit-shared-theme/item'; // Used for Group titles and Item descriptions

exports.themeNamespace = themeNamespace;
var compactSmallFontSize = 10;
exports.compactSmallFontSize = compactSmallFontSize;
var compactLineHeight = 1.2;
exports.compactLineHeight = compactLineHeight;
var gridSize = (0, _theme.gridSize)();
exports.gridSize = gridSize;
var defaultTheme = {
  afterItemSpacing: {
    compact: (0, _theme.gridSize)(),
    default: (0, _theme.gridSize)()
  },
  beforeItemSpacing: {
    compact: (0, _theme.gridSize)(),
    default: (0, _theme.gridSize)()
  },
  borderRadius: (0, _theme.borderRadius)(),
  focus: {
    outline: _theme.colors.B100
  },
  height: {
    compact: 0,
    default: 0
  },
  padding: {
    default: {
      bottom: (0, _theme.gridSize)() / 2,
      left: (0, _theme.gridSize)() / 2,
      right: (0, _theme.gridSize)() / 2,
      top: (0, _theme.gridSize)() / 2
    },
    compact: {
      bottom: (0, _theme.gridSize)(),
      left: (0, _theme.gridSize)(),
      right: (0, _theme.gridSize)(),
      top: (0, _theme.gridSize)()
    }
  },
  default: {
    background: _theme.colors.background,
    text: _theme.colors.text,
    secondaryText: _theme.colors.N200
  },
  selected: {
    background: _theme.colors.backgroundActive,
    text: _theme.colors.N500,
    secondaryText: _theme.colors.N200
  },
  active: {
    background: _theme.colors.backgroundActive,
    text: _theme.colors.textActive,
    secondaryText: _theme.colors.N200
  },
  hover: {
    background: _theme.colors.backgroundHover,
    text: _theme.colors.textHover,
    secondaryText: _theme.colors.N200
  },
  disabled: {
    background: _theme.colors.N0,
    text: _theme.colors.N200,
    secondaryText: _theme.colors.N200
  },
  // same as hover in this case
  dragging: {
    background: _theme.colors.N20,
    text: _theme.colors.N800,
    secondaryText: _theme.colors.N200
  }
};
exports.defaultTheme = defaultTheme;

var isValidCssValue = function isValidCssValue(value) {
  return value !== undefined && value !== null && value !== '';
}; // Returns the theme that contains the requested theme key(s), preferring the user-supplied
// theme if it is provided.


var themeWithKeys = function themeWithKeys(maybeTheme, key, parentKey) {
  if (parentKey) {
    return maybeTheme && maybeTheme[parentKey] && isValidCssValue(maybeTheme[parentKey][key]) ? maybeTheme : defaultTheme;
  }

  return maybeTheme && isValidCssValue(maybeTheme[key]) ? maybeTheme : defaultTheme;
}; // Returns the theme value for the requested key(s), falling back to the default theme if the
// user-supplied theme doesn't exist or doesn't contain the requested key(s)


exports.themeWithKeys = themeWithKeys;

var getThemeStyle = function getThemeStyle(maybeTheme, key, parentKey) {
  var theme = themeWithKeys(maybeTheme, key, parentKey);
  return parentKey ? theme[parentKey][key] : theme[key];
};

exports.getThemeStyle = getThemeStyle;
